[inductor] Implement bucketize() for dependencies.py

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105102

dependencies.py is used for tracking reads and writes, which is used for identifying dependencies between buffers: i.e. if buffer X reads buffer Y, then X depends on Y. ops.bucketize() reads from an offsets tensor, so we should track it in dependencies.py to correctly track dependencies. Since bucketize performs a binary search over the offsets tensor, the dependency is marked as a StarDep to indicate that the entire tensor is needed.

Use case: we find that jagged tensor dense_to_jagged ops - which use bucketize() to map jagged indices to dense indices - perform better if the bucketize() kernel is separated from the gather kernel. Previously, because bucketize() wasn't marked as reading anything, it would just get inlined.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

Differential Revision: [D47422704](https://our.internmc.facebook.com/intern/diff/D47422704)

davidberard98 committed:
[inductor] Implement bucketize() for dependencies.py

dependencies.py is used for tracking reads and writes, which is used for identifying dependencies between buffers: i.e. if buffer X reads buffer Y, then X depends on Y. ops.bucketize() reads from an offsets tensor, so we should track it in dependencies.py to correctly track dependencies. Since bucketize performs a binary search over the offsets tensor, the dependency is marked as a StarDep to indicate that the entire tensor is needed.

Use case: we find that jagged tensor dense_to_jagged ops - which use bucketize() to map jagged indices to dense indices - perform better if the bucketize() kernel is separated from the gather kernel. Previously, because bucketize() wasn't marked as reading anything, it would just get inlined.

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47422704).

davidberard98 committed:
Update on "[inductor] Implement bucketize() for dependencies.py"


dependencies.py is used for tracking reads and writes, which is used for identifying dependencies between buffers: i.e. if buffer X reads buffer Y, then X depends on Y. ops.bucketize() reads from an offsets tensor, so we should track it in dependencies.py to correctly track dependencies. Since bucketize performs a binary search over the offsets tensor, the dependency is marked as a StarDep to indicate that the entire tensor is needed.

Use case: we find that jagged tensor dense_to_jagged ops - which use bucketize() to map jagged indices to dense indices - perform better if the bucketize() kernel is separated from the gather kernel. Previously, because bucketize() wasn't marked as reading anything, it would just get inlined.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

Differential Revision: [D47422704](https://our.internmc.facebook.com/intern/diff/D47422704)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47422704).

aakhundov commented on torch/_inductor/codegen/triton.py:
```diff
@@ -2409,7 +2410,7 @@ def candidate_tilings(node):
             dep
             for dep in itertools.chain(rw.reads, rw.writes)
-            if dep.name not in V.graph.removed_buffers
+            if dep.name not in V.graph.removed_buffers and isinstance(dep, MemoryDep)
```
Curious, why do we want to ignore other `dep` types than `MemoryDep` here?

davidberard98 replied:
Practically, because later we use `dep.index` which doesn't exist on StarDeps. Conceptually my understanding is that StarDep accesses also won't contribute suggestions for tiling since they don't have indexing information.

Added a comment

davidberard98 committed:
Update on "[inductor] Implement bucketize() for dependencies.py"


dependencies.py is used for tracking reads and writes, which is used for identifying dependencies between buffers: i.e. if buffer X reads buffer Y, then X depends on Y. ops.bucketize() reads from an offsets tensor, so we should track it in dependencies.py to correctly track dependencies. Since bucketize performs a binary search over the offsets tensor, the dependency is marked as a StarDep to indicate that the entire tensor is needed.

Use case: we find that jagged tensor dense_to_jagged ops - which use bucketize() to map jagged indices to dense indices - perform better if the bucketize() kernel is separated from the gather kernel. Previously, because bucketize() wasn't marked as reading anything, it would just get inlined.

cc voznesenskym penguinwu @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

Differential Revision: [D47422704](https://our.internmc.facebook.com/intern/diff/D47422704)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47422704).

eellison commented on torch/_inductor/ir.py:
```diff
@@ -2219,11 +2219,13 @@ def get_fill_order(self):
                     r.index, {v: sympy.Integer(0) for v in reduction_vars if v != 0}
                 )
                 for r in reads
+                if isinstance(r, dependencies.MemoryDep)
```
What prompted you to make this change, just curious ? I guess as above you got an error because `StarDeps` doesn't have `index`.

davidberard98 replied:
Yeah, that's why. I guess previously, ComputedBuffers never had StarDeps.

davidberard98 replied:
yeah just what you said, trying to get r.index.

eellison commented on torch/_inductor/ir.py:
```diff
@@ -2219,11 +2219,13 @@ def get_fill_order(self):
                     r.index, {v: sympy.Integer(0) for v in reduction_vars if v != 0}
                 )
                 for r in reads
+                if isinstance(r, dependencies.MemoryDep)
```
nit: maybe assert that that all the reads are `MemoryDep` or `StarDep` ?

davidberard98 replied:
yeah definitely a good idea

eellison commented on torch/_inductor/dependencies.py:
```diff
@@ -239,6 +241,17 @@ def index_expr(self, index: sympy.Expr, dtype) -> str:
+        indexing_dtype: torch.dtype,
+        right: bool,
+    ):
+        self._reads.add(StarDep(offsets_name))
```
Why not :

`self._writes.add(MemoryDep(offsets_name, *self.canonicalize(offsets_size)))` ?

davidberard98 replied:
AFAIK, the `index` field is intended to refer to the indices that are accessed during the read. In this case, the binary search needs the raw `offsets_ptr` and then it will binary search over the offsets, so it actually does need the entire tensor and we don't actually know the indices it will need, because that's data-dependent.

offsets_size also isn't the index we're accessing, it's the full size of the tensor.

but I could be wrong about this interpretation, if so LMK

eellison replied:
Should we just add the dependency as the entire tensor then ?

eellison reviewed (commented):
nice! one question about StarDep vs MemDep

davidberard98 committed:
Update on "[inductor] Implement bucketize() for dependencies.py"


dependencies.py is used for tracking reads and writes, which is used for identifying dependencies between buffers: i.e. if buffer X reads buffer Y, then X depends on Y. ops.bucketize() reads from an offsets tensor, so we should track it in dependencies.py to correctly track dependencies. Since bucketize performs a binary search over the offsets tensor, the dependency is marked as a StarDep to indicate that the entire tensor is needed.

Use case: we find that jagged tensor dense_to_jagged ops - which use bucketize() to map jagged indices to dense indices - perform better if the bucketize() kernel is separated from the gather kernel. Previously, because bucketize() wasn't marked as reading anything, it would just get inlined.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

Differential Revision: [D47422704](https://our.internmc.facebook.com/intern/diff/D47422704)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47422704).

davidberard98 committed:
lint on "[inductor] Implement bucketize() for dependencies.py"


dependencies.py is used for tracking reads and writes, which is used for identifying dependencies between buffers: i.e. if buffer X reads buffer Y, then X depends on Y. ops.bucketize() reads from an offsets tensor, so we should track it in dependencies.py to correctly track dependencies. Since bucketize performs a binary search over the offsets tensor, the dependency is marked as a StarDep to indicate that the entire tensor is needed.

Use case: we find that jagged tensor dense_to_jagged ops - which use bucketize() to map jagged indices to dense indices - perform better if the bucketize() kernel is separated from the gather kernel. Previously, because bucketize() wasn't marked as reading anything, it would just get inlined.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

Differential Revision: [D47422704](https://our.internmc.facebook.com/intern/diff/D47422704)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47422704).

eellison commented on test/inductor/test_dependencies.py:
```diff
@@ -0,0 +1,64 @@
+        super().tearDown()
+
+    @unittest.skipIf(not HAS_CUDA, "CUDA-only test")
+    def test_bucketize_dependencies(self):
```
maybe worth adding a test that tests the fusion you want to occur does occur, with `run_and_get_code`

davidberard98 replied:
good idea - I added a `self.assertEqual(torch._inductor.metrics.generated_kernel_count, 1)` into a bucketize -> add test in `test_torchinductor.py`

davidberard98 committed:
Update on "[inductor] Implement bucketize() for dependencies.py"


dependencies.py is used for tracking reads and writes, which is used for identifying dependencies between buffers: i.e. if buffer X reads buffer Y, then X depends on Y. ops.bucketize() reads from an offsets tensor, so we should track it in dependencies.py to correctly track dependencies. Since bucketize performs a binary search over the offsets tensor, the dependency is marked as a StarDep to indicate that the entire tensor is needed.

Use case: we find that jagged tensor dense_to_jagged ops - which use bucketize() to map jagged indices to dense indices - perform better if the bucketize() kernel is separated from the gather kernel. Previously, because bucketize() wasn't marked as reading anything, it would just get inlined.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

Differential Revision: [D47422704](https://our.internmc.facebook.com/intern/diff/D47422704)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47422704).

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot merge

voznesenskym commented:
@pytorchbot revert

voznesenskym commented:
@pytorchbot revert -m "Regresses perf all over the place" -c nosignal

davidberard98 committed:
Update on "[inductor] Implement bucketize() for dependencies.py"


dependencies.py is used for tracking reads and writes, which is used for identifying dependencies between buffers: i.e. if buffer X reads buffer Y, then X depends on Y. ops.bucketize() reads from an offsets tensor, so we should track it in dependencies.py to correctly track dependencies. Since bucketize performs a binary search over the offsets tensor, the dependency is marked as a StarDep to indicate that the entire tensor is needed.

Use case: we find that jagged tensor dense_to_jagged ops - which use bucketize() to map jagged indices to dense indices - perform better if the bucketize() kernel is separated from the gather kernel. Previously, because bucketize() wasn't marked as reading anything, it would just get inlined.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D47422704](https://our.internmc.facebook.com/intern/diff/D47422704)

[ghstack-poisoned]

davidberard98 committed:
fix usage of itertools.chain on "[inductor] Implement bucketize() for dependencies.py"


dependencies.py is used for tracking reads and writes, which is used for identifying dependencies between buffers: i.e. if buffer X reads buffer Y, then X depends on Y. ops.bucketize() reads from an offsets tensor, so we should track it in dependencies.py to correctly track dependencies. Since bucketize performs a binary search over the offsets tensor, the dependency is marked as a StarDep to indicate that the entire tensor is needed.

Use case: we find that jagged tensor dense_to_jagged ops - which use bucketize() to map jagged indices to dense indices - perform better if the bucketize() kernel is separated from the gather kernel. Previously, because bucketize() wasn't marked as reading anything, it would just get inlined.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

Differential Revision: [D47422704](https://our.internmc.facebook.com/intern/diff/D47422704)

[ghstack-poisoned]

davidberard98 commented:
whoops, looks like I was reusing the `itertools.chain(rw.reads, rw.writes)` in triton.py - on the second use the iterator was already exhausted so the deps list would be empty.

fixed by constructing the `itertools.chain` for each of its uses.

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
fyi, verified no regression on the dashboard from the fixed version of this PR:

https://hud.pytorch.org/benchmark/compilers?startTime=Tue%2C%2011%20Jul%202023%2017%3A07%3A27%20GMT&stopTime=Tue%2C%2018%20Jul%202023%2017%3A07%3A27%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/davidberard98/206/orig&lCommit=719b8524a3442f4341782233284b81f4ee5b2dfa&rBranch=davidberard98-206-base&rCommit=e68cf02420b852c3bf278fe9d1f47d6b02bae0f0

<img width="1255" alt="Screenshot 2023-07-18 at 10 08 10 AM" src="https://github.com/pytorch/pytorch/assets/5067123/2e572922-54fa-453e-8421-28cffaade9a6">
